IncludeToolbox/Options/FormatterOptionsPage.cs

```diff
@@ -19,16 +19,29 @@ public enum PathMode
             Shortest,
             Shortest_AvoidUpSteps,
             Absolute,
+            Absolute_FromParentDirWithFile
         }
         [Category("Path")]
         [DisplayName("Mode")]
         [Description("Changes the path mode to the given pattern.")]
         public PathMode PathFormat { get; set; } = PathMode.Shortest_AvoidUpSteps;
 
         [Category("Path")]
+        [DisplayName("Filename in parent directory for absolute include paths")]
+        [Description("The Absolute_FromParentDirWithFile mode will look for this file in parent directories and make include paths absolute from its location.")]
+        public string FromParentDirWithFile { get; set; } = "build.root";
+
+        public enum IgnoreFileRelativeMode
```

Comment:
Originally this was intended to suppress the "natural include path" even if that one would be better according to the policy in PathMode.
You're right, this is all not too intuitive. But I think with your extension we completely lost the meaning of "ignore" (never a good description of what was going on). So maybe this would be more concise?

enum AllowRelativePathMode
{
Always, // All relative paths are taken into consideration.
Never, // No relative path is taken into consideration.
OnlyInSameDirectory, // Relative paths are only allowed if they are in the very same directory than the including file.
OnlyInSameOrSubDirectory // Relative paths are allowed if they are in the same directory than the including file or any subdirectory.
}

Reply:
I think it might be a better idea to do something different. What do you think about this:

### Remove Always/Never
- To _resolve_ the full path to the include, we should automatically use the including file's directory if the delimiters are quotes (https://msdn.microsoft.com/en-us/library/36k2cdd4.aspx)
  - This would go well with what I'd like to do next: Properly handle system includes (adding an `Auto` to `DelimiterFormatting` and handling their formatting specially (there's a TODO for something like this already))
- This mode will instead optionally **force** use of paths relative to the including file's directory, overriding `PathFormat` if set
  - I really need this behavior -- I want to `ForceRelativeToParentDirWithFile` **unless** the include is a system one or in the same directory as the including file
- If this mode is disabled, actually formatting include paths as relative to anything else will be left to the `PathFormat`

Basically I'm envisioning these two being like this:

    public enum PathMode
    {
        Unchanged,
        Shortest,               // Maybe for these we can check to see if there's a shorter one than the
        Shortest_AvoidUpSteps,  // force-file-relative mode instead of overriding them entirely 
        Absolute,
        ForceRelativeToParentDirWithFile
    }
    
    public enum ForceFileRelativePathMode
    {
        Disabled,               // Off
        IfInSameDirectory,      // DEFAULT. Includes found in the directory containing the including
                                // file are always made relative to it. Does not include subdirectories.
        IfInSameOrSubDirectory  // Includes found in the directory containing the including file are
                                // always made relative to it. Includes subdirectories.
    }

Reply:
I think you're definitely right on the resolve strategy!  Didn't know about "2" for '"' delimiters. Taking this into account would be hard ...

I also agree on most of the rest as well but, as you're pointing out neither `Shortest`, `Shortest_AvoidUpSteps`, nor `ForceRelativeToParentDirWithFile` shouldn't be too much intimidated by `ForceFileRelativePathMode`. That being said we don't really force relative paths. Also, your suggested setup does not make it clear what you need to do to allow relative paths going outside of the source dir, e.g. "../header.h".
It seems all our modes strive "use the shortest out of the following", only that the list to choose from is differently every time. Looking at things from that perspective we would end up with various bit flags that govern which path are legal for either <> or "". But since this doesn't work with the UI, it might translate to this:

```
public enum PathFormattingMode
{
    Unchanged,
    Shortest, // default
    ForceAbsolute // absolute, no matter what. for debugging mainly
}
// Relative paths are never considered for includes using <>
public enum PathPool_FileRelativePath
{
    DontConsider, // pretend relative paths are not possible
    Consider, // add relative path like a normal include path
    ConsiderOnlyIfInSameDirectory, // Allow relative only if in same directory
    ConsiderOnlyIfInSameOrSubDirectory, // Allow relative only if in same directory or subdirectory
}
public enum PathPool_ProjectIncludePath
{
   AlwaysConsider // default
   ConsiderOnlyForAngled, // allows you to use enforce using relative or RelativeToParentDirWithFile for quoted ones
}
public enum PathPool_RelativeToParentDirWithFile
{
   DontConsider,
   AlwaysConsider,
   DontConsiderForAngled
}
public bool PathPool_AllowUpSteps = false; // If true allow "../" in directories
```
If I got everything right that would mean that your new desired mode is configured as `Shortest`, `PathPool_FileRelativePath==ConsiderOnlyIfInSameDirectory` , `PathPool_ProjectIncludePath==ConsiderOnlyForAngled`, `PathPool_RelativeToParentDirWithFile==DontConsiderForAngled`, `PathPool_AllowUpSteps==false`
Yeah I know, is this _really_ better? Just trying to find flexible and easy to understand framework that fits everything :-/

About the delimiter thing: Having auto mode sounds great! Just don't rely on any semantic for existing/targeted delimiters, I worked in projects where all delimiters were angle brackets by convention.

Reply:
With respect to `Shortest` and my workflow, I actually want "as long as it needs to be, starting from the root of our repository", but _only for files not in the current project_ (and not for system files, of course).

I've already seen a couple of cases in our codebase where files that _are_ in the current directory had longer relative paths that my current implementation fixed appropriately.

As an example, consider:

* RepoDir/
  * build.root
  * Src/
    * MainProj/
      * MainProj.cpp
      * MainProj.h
      * MainUtility.h
      * stdafx.h
    * ProjA/
      * A_1.h
      * A_2.h
    * ProjB/
      * B_1.h
      * B_2.h

Group regexes:

```
(?i)stdafx\.h(?-i)                          // PCH
"(?i)$(currentFilename)\.(h|hpp|hxx)(?-i)"  // Corresponding header
<[^.]*\.                                    // C system includes
<                                           // C++ system includes
(\\|/)                                      // Other projects' headers
                                            // The rest (this project's headers)
```

Original:

```c++
// MainProj.cpp
// AdditionalIncludeDirs="../ProjA;../ProjB;$(BuildRootDirMacro);..."

#include "stdafx.h"

#include "A_2.h"
#include "B_1.h"
#include "B_2.h"
#include "MainProj.h"
#include "../Src/ProjA/A_1.h"
#include "../MainProj/MainUtility.h"

#include <vector>
#include <windows.h>
#include <string>
```

Formatted:

```c++
// MainProj.cpp
// AdditionalIncludeDirs="../ProjA;../ProjB;$(BuildRootDirMacro);..."

#include "stdafx.h"

#include "MainProj.h"

#include <windows.h>

#include <vector>
#include <string>

#include "Src/ProjA/A_1.h"
#include "Src/ProjA/A_2.h"
#include "Src/ProjB/B_1.h"
#include "Src/ProjB/B_2.h"

#include "MainUtility.h"
```

If you're curious, that grouping/order comes from the [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes), which is what our style guide is based on. My reason for wanting those long, rooted paths are for a few reasons:

1. Standardization
2. Easy searching
3. Easy parsing with other tools! If you don't have to worry about digging settings out of the project files and property sheets, parsing includes becomes much easer.

Reply:
In terms of "paths going outside of the source dir", I'll admit that's not a case that affects me personally so I wasn't terribly concerned about it, but I think failing to fit other strategies and falling back on one of the `Shortest` options or even `Unchanged` may cover it.

Reply:
I really think the [wiki page](https://github.com/dakotahawkins/IncludeToolbox/wiki/Include-Path-Formatting-(Design)) would be good to sort this out, since my intention was to take a step back from the implementation and address what kind of general behavior we'd like to support.

Also, with so many options that affect one another, the possible behaviors really scale. It might be OK if there are some combinations of them that don't make sense and produce bad/silly results as long as those combinations aren't easy to accidentally have and the "good" combinations allow lots of kinds of results that people might actually want or need.